### PR TITLE
SR: Extract Avro schema with reflection

### LINKED ISSUE #1041: SR: Extract Avro schema from structured data
When auto-registering is enabled we need to extract the schema from a record to register it in the schema registry. The goal of this issue is to create a function that traverses the map and creates an Avro schema from it. Similar to how we extract the schema in the [SDK](https://github.com/ConduitIO/conduit-connector-sdk/blob/a64baed95dbc0f1e92df7ad12f3aa0c7a8ceee50/kafkaconnect/reflect.go#L25-L27).

### PR DESCRIPTION
**Pull request by lovromazgon**
### Description

This PR adds a function that traverses a struct (normally structured data) and produces an Avro schema from it. A good chunk of the code was [taken from the SDK](https://github.com/ConduitIO/conduit-connector-sdk/blob/main/kafkaconnect/reflect.go). Next time we add a format it would be advisable to extract the common code.

This approach still contains the drawback that it could produce a different schema if the provided value contains a `nil` value (we have no type info in that case).

Fixes #1041 

### Quick checks:

- [X] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [X] There is no other [pull request](https://github.com/ConduitIO/conduit/pulls) for the same update/change.
- [X] I have written unit tests.
- [X] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

File: pkg/processor/schemaregistry/avro/schema.go
Comment (lovromazgon) [reply pkg/processor/schemaregistry/avro/schema.go:100]: Good point, currently we make an assumption they are all the same type. I can try to take the type info from the slice type, although the issue is that I'm not sure if the type info of slices is preserved when the connector sends us the data. Also, if we are receiving raw data (e.g. JSON) and are unmarshaling it into structured data, we will almost certainly lack the type info for slices.

To sum up - the question is if we see `[]any{1,2,3}`, should we assume that the type is always `int`, even though theoretically the slice could contain any type? 🤔

Comment (hariso) [reply pkg/processor/schemaregistry/avro/schema.go:100]: Can you remind me, how would schema updates work? I'm wondering if we can use a union type here and update it if needed. E.g. if we have `[]any{"hi", 123}` we have a union type with the string and int types. A new type might be added in the future, in which case we update the schema (if possible).

Comment (lovromazgon) [reply pkg/processor/schemaregistry/avro/schema.go:100]: We don't start with an existing schema, we simply extract it from the data that we have in the current record. There is no update in that sense. We try to register the extracted schema and either get back an ID of an existing schema (ideally it was already cached, so we didn't even send a request), or the new updated schema gets stored in the registry and we get back a new ID.

Comment (hariso) [reply pkg/processor/schemaregistry/avro/schema.go:151]: @lovromazgon so for the new, updated, schema -- how is it connected to the previous one, for the records from the same source? Thinking aloud, a source _may_ return homogenous data, so we can't assume that all records belong to the same schema, so it has to be a new schema always?

Comment (lovromazgon) [reply pkg/processor/schemaregistry/avro/schema.go:151]: The processor will have an option to configure the subject for the extracted schema, the schema registry figures out the version (if it already contains the exact schema under that subject it uses that, otherwise creates a new version). The subject can be static or taken from the record (e.g. from metadata), so we should be able to handle different records.

File: pkg/processor/schemaregistry/avro/schema.go
Comment (hariso) [inline pkg/processor/schemaregistry/avro/schema.go:253]: I'm probably missing something obvious, but why are uint64 and unit not supported, but unit32 for example is? Also, what's the reason for `complex` not being supported?

Comment (lovromazgon) [reply pkg/processor/schemaregistry/avro/schema.go]: There's no equivalent type in Avro for these types, the Avro library doesn't support them either. We could add support by marshaling them as [`fixed`](https://stackoverflow.com/a/75242092) or simply `bytes`, but we wouldn't know the original type when unmarshaling, we'd get a `[]byte` on the other side. Maybe custom types could be used to forward the type info.

Given that it's not trivial to add the support and given that they are not _that_ common I skipped them and I plan to add a note to the limitations of the processor.

Comment (hariso) [reply pkg/processor/schemaregistry/avro/schema.go]: Makes sense to me. If not already, it would be good to have a word or two about that in the code.

File: pkg/processor/schemaregistry/avro/schema_test.go
Comment (hariso) [inline pkg/processor/schemaregistry/avro/schema_test.go:357]: Nitpick: it would be nice if we could check the error itself too (i.e. that it didn't fail for another reason).

File: pkg/processor/schemaregistry/avro/schema.go
Comment (lovromazgon) [reply pkg/processor/schemaregistry/avro/schema.go]: `IsValid` will be `false` when `reflect.Value` is the zero value (e.g. `reflect.Value{}`). We sometimes pass down the zero value, for example when we are trying to extract the schema of a slice with a concrete type (e.g. `[]int`). in that case we ignore the actual values in the schema and simply rely on the type information of the slice. If that type is a struct (e.g. `[]MyStruct`) we will further traverse the fields and have no value to send alongside the field, so it can happen that the value is invalid.

File: pkg/processor/schemaregistry/avro/schema.go
Comment (lovromazgon) [reply pkg/processor/schemaregistry/avro/schema.go:22]: I'm waiting on a response from the maintainer on two PRs there, we only need one of them though. Once it's merged we can switch back to the original lib, we will have to adjust our code just a bit (the API changed, functionality should be the same though). I'm keeping an eye on that and will do that once it's done 👍

File: pkg/processor/schemaregistry/avro/reflect.go
Comment (lovromazgon) [reply pkg/processor/schemaregistry/avro/reflect.go:80]: If we didn't do that, a field could be lost, it would be like it's not even there. For example, when we get a value like `map[string]any{"foo":nil}` we have no idea what `nil` could be, so we fall back to the nullable string.

I'll mention it in code 👍

File: pkg/processor/schemaregistry/avro/schema.go
Comment (lovromazgon) [reply pkg/processor/schemaregistry/avro/schema.go:31]: I added a comment 🙂 It's because of how unions are marshled / unmarshaled in the lib we're using (unfortunately other libs work the same, otherwise I would have switched).

File: pkg/processor/schemaregistry/avro/reflect.go
Comment (lovromazgon) [reply pkg/processor/schemaregistry/avro/reflect.go]: Also I moved `typesSet` into `duplicate` 👍

File: pkg/processor/schemaregistry/avro/reflect.go
Comment (lovromazgon) [reply pkg/processor/schemaregistry/avro/reflect.go]: I added comments on all extractX methods, including `extractPointer`.

File: pkg/processor/schemaregistry/avro/schema_test.go
Comment (hariso) [reply pkg/processor/schemaregistry/avro/schema_test.go:139]: It would be nice having that in a comment, just so it reminds us of that limitation in the libs we're using.

File: pkg/processor/schemaregistry/avro/union.go
Comment (hariso) [inline pkg/processor/schemaregistry/avro/union.go:61]: > traverses the value using the schema

"the schema" is the one used in NewUnionResolver I guess? Since in UnionResolver itself we have no reference to a schema.

**Review by hariso — APPROVED** [review]
LGTM!

File: pkg/processor/schemaregistry/avro/union.go
Comment (lovromazgon) [reply pkg/processor/schemaregistry/avro/union.go:61]: Yes, we essentially pre-calculate the paths we are interested in, so we don't traverse the whole schema every time, but only when the union resolver is first created.

File: pkg/processor/schemaregistry/avro/schema.go
Comment (lovromazgon) [reply pkg/processor/schemaregistry/avro/schema.go]: I added a comment to `extractor.Extract` about this.

**Pull request merged**
